ts/user-picker/user-picker.controller.ts

```diff
@@ -0,0 +1,342 @@
+					angular.isDefined(newValue[1]) && newValue[1].length > 0 &&
+					newValue[0] != oldValue[0] && !_.isEqual(newValue[1], oldValue[1])) {
+					this.resetUsers();
+					this.refresh();
+				}
+			});
+
+			this.$scope.find = (search: string): void => {
+				this.resetUsers();
+				this.refresh(search);
+			};
+
+			this.$scope.loadMore = (): void => {
+				if (this.$scope.displayAllUsers === undefined || this.$scope.displayAllUsers === false) {
+					this.$scope.lastPagingOffset += MAGIC_PAGING;
+					this.$scope.loadingMore = true;
+					this.refresh().then(() => { this.$scope.loadingMore = false; });
+				}
+			};
+
+			this.$scope.onSelectedUserChanged = (user: IUserLookup): void => {
+				this.setViewValue(user);
+				if (!!this.$scope.onSelect()) {
+					this.$scope.onSelect();
+				}
+			};
+
+			this.$scope.onSelectedUsersChanged = (): void => {
+				this.setViewValue(this.$scope.selectedUsers);
+				if (!!this.$scope.onSelect()) {
+					this.$scope.onSelect();
+				}
+			};
+
+			this.$scope.onSelectedUserRemoved = (): void => {
+				this.setViewValue(this.$scope.selectedUsers);
+				if (!!this.$scope.onRemove()) {
+					this.$scope.onRemove();
+				}
+			};
+		}
+
+		/**
+		 * Fetches most of the additional properties given in the attributes of the directive (custom-info, customInfoAsync, etc).
+		 * Also handles homonyms.
+		 */
+		private tidyUp(users: IUserLookup[]): ng.IPromise<IUserLookup[]> {
+			let promises = new Array<ng.IPromise<any>>();
+			let customInfoDico: { [userId: number]: ng.IPromise<string> } = {};
+			let homonymsDico = {};
+
+			_.each(users, (user: IUserLookup) => {
+				user.hasLeft = !!user.dtContractEnd && moment(user.dtContractEnd).isBefore(moment().startOf("day"));
+			});
+
+			if (this.$scope.customFilter) {
+				users = _.filter(users, (user: IUserLookup) => { return this.$scope.customFilter(user); });
+			}
+			if (!!this.$scope.customInfo) {
+				_.each(users, (user: IUserLookup) => {
+					user.info = this.$scope.customInfo(user);
+				});
+			}
+			if (!!this.$scope.customInfoAsync) {
+				_.each(users, (user: IUserLookup) => {
+					customInfoDico[user.id.toString()] = promises.push(this.$scope.customInfoAsync(user)) - 1;
+				});
+			}
+
+			let homonyms = this.userPickerService.getHomonyms(users);
+			if (!!homonyms && homonyms.length > 0) {
+				let properties = !!this.$scope.homonymsProperties && this.$scope.homonymsProperties.length > 0 ?
+					this.$scope.homonymsProperties : DEFAULT_HOMONYMS_PROPERTIES;
+				_.each(homonyms, (user: IUserLookup) => {
+					homonymsDico[user.id] = promises.push(this.userPickerService.getAdditionalProperties(user, properties)) - 1;
+				});
+			}
+
+			return this.$q.all(promises).then((values: any[]) => {
+				if (!!homonyms && homonyms.length > 0) {
+					_.each(users, (user: IUserLookup) => {
+						if (angular.isDefined(homonymsDico[user.id])) {
+							user.additionalProperties = values[homonymsDico[user.id.toString()]];
```

Comment:
Same: I don't think you need to do `.toString()` since you expect the key to be a number.